### DISCUSSION TITLE
Variable depends on faces

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26106] [Q&A General]
Hello everyone,

I am currently working on a project where I need to compute the flow on specific faces of the mesh. However, it seems that in MOOSE, the solution is primarily stored at quadrature points (qp) or nodal locations.I am wondering if there is a way to define a variable related to faces, allowing me to store information about each face, such as its area, etc. I am open to using auxvariables or postprocessors to achieve this.

I have explored InternalSideUserObject and its subclass ElementSidesL2Norm. However, it appears that ElementSidesL2Norm performs a cumulative calculation over all faces, and other subclasses seem to perform cumulative calculations on interfaces. What I need is an individual calculation for each face, not a cumulative result.

Moreover, I have questions about how InternalSideUserObject traverses the mesh for computations. I noticed that both elem and side references in InternalSideUserObject are const.
```
/// pointer to the current element object
  const Elem * const & _current_elem;

  /// the volume of the current element
  const Real & _current_elem_volume;

  /// current side of the current element
  const unsigned int & _current_side;
```
By tracing the code, I suspect that MOOSE performs the mesh traversal through the loop in ThreadedElementLoopBase<RangeType>::operator()(const RangeType & range, bool bypass_threading), which repeatedly calls ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side).However, I couldn't find in ComputeUserObjectsThread::onInternalSide() how MOOSE invokes the relevant InternalSideUserObject for computation based on the input parameters elem and side. Instead, it seems MOOSE repeatedly calls a batch of InternalSideUserObject objects.
```
void
ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side)
{
  // Pointer to the neighbor we are currently working on.
  const Elem * neighbor = elem->neighbor_ptr(side);

  // Get the global id of the element and the neighbor
  const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();

  if (_internal_side_objs.size() == 0 && _domain_objs.size() == 0)
    return;
  if (!((neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) ||
        (neighbor->level() < elem->level())))
    return;

  _fe_problem.prepareFace(elem, _tid);
  _fe_problem.reinitNeighbor(elem, side, _tid);

  // Set up Sentinels so that, even if one of the reinitMaterialsXXX() calls throws, we
  // still remember to swap back during stack unwinding.
  SwapBackSentinel face_sentinel(_fe_problem, &FEProblem::swapBackMaterialsFace, _tid);
  _fe_problem.reinitMaterialsFace(elem->subdomain_id(), _tid);

  SwapBackSentinel neighbor_sentinel(_fe_problem, &FEProblem::swapBackMaterialsNeighbor, _tid);
  _fe_problem.reinitMaterialsNeighbor(neighbor->subdomain_id(), _tid);

  for (const auto & uo : _internal_side_objs)
    if (!uo->blockRestricted() || uo->hasBlocks(neighbor->subdomain_id()))
      uo->execute();

  for (auto & uo : _domain_objs)
    if (!uo->blockRestricted() || uo->hasBlocks(neighbor->subdomain_id()))
    {
      uo->preExecuteOnInternalSide();
      uo->executeOnInternalSide();
    }
}
```
Any guidance or insights on how to define a face-related variable and efficiently perform calculations on each face individually in MOOSE would be greatly appreciated.

Thank you!

**Comment by GiudGiud** [comment]
Hello

We have not worked with face variables yet in MOOSE but they were added to libmesh.
@roystgnr or @lindsayad do you have a roadmap for face variable use in MOOSE somewhere? From the discussions with Yaqi

Guillaume

**Reply by lindsayad** [reply]
@roystgnr has a demonstration of their use in https://github.com/idaholab/moose/blob/next/test/tests/dgkernels/hfem/hfem_jacobian.i

**Reply by hityyds** [reply]
Thank you very much for your and your friend's responses. I will carefully study this demonstration.

**Reply by roystgnr** [reply]
The one thing that worries me is: "specific faces of the mesh".  The `SIDE_HIERARCHIC` variables will have DoFs on *every* face of every element in the subdomains where they're defined.  There are ways around this (if you have an Auxilliary system variable you might just ignore the faces you don't care about; if you have a Nonlinear system variable you might just add a constraint to those faces), but if you really only care about a small subset of faces then the solution may be to add lower-D elements (e.g. a Quad9 attached to the side of a Hex27 on those specific face.
